### ISSUE TITLE
Lens flares are drawn through objects

### ISSUE THREAD

**Issue by mmillar-bolis** [fgsfdsfgs/perfect_dark#119]
At least on Win10 with Nvidia drivers, it appears that lens flares can be seen through some dynamic objects such as the player hands, the flying saucer at the end of `Area 51: Escape`, and the rotating pillar in `Skedar Ruins`.

Area 51:
![lights00](https://github.com/fgsfdsfgs/perfect_dark/assets/62162171/f1d15640-c8f2-41d3-9b0d-2908755b6d38)

Skedar Ruins:
![lights01](https://github.com/fgsfdsfgs/perfect_dark/assets/62162171/e14e002a-f028-4ce0-972a-b33717f9e9c5)

The small green and blue flares also do this:
![lights02](https://github.com/fgsfdsfgs/perfect_dark/assets/62162171/a7670440-632c-4f8d-ad3e-16dd4fdb7589)

**Comment by fgsfdsfgs** [comment]
This is actually not related to the GPU at all.
In the port the depth testing for flares has been replaced with LOS tracing because doing what the game does on the N64 is expensive on modern platforms and this issue is the result of that. While I will fix the flares rendering through some world objects, I probably won't fix them rendering through your gun.

**Comment by ghost** [comment]
![fan](https://github.com/fgsfdsfgs/perfect_dark/assets/130617241/a1f9c4b7-6c8d-4475-9ba4-5c1c5ef08922)

Flares rendering through objects should be fixed since there there's a room in Defection showing how objects can covers flares with this rotating fan.

**Comment by mmillar-bolis** [comment]
> what the game does on the N64 is expensive on modern platforms and this issue is the result of that. While I will fix the flares rendering through some world objects, I probably won't fix them rendering through your gun.

That certainly makes sense, and it's not terribly noticeable on the hand models during natural gameplay.

**Comment by ghost** [comment]
Yep, depth is expensive. In GLideN64, special shaders are done to work out per pixel depth, and on some systems it can be a major performance hit.  Including one of the N64's weird depth compare modes, and it also uses special mipmap shaders for LOD emulation.  Makes sense to hack around it, since basically you got a full decomp and from there you can just improve on the game code to be more efficient on PCs/Switches/etc.

**Comment by fgsfdsfgs** [comment]
The game already traces gunshots through (almost) all visible geometry, including prop and character models, so leveraging that seemed like a good idea.
I haven't worked out some quirks of how the shot trace works yet, so for now it's doing a simpler line-of-sight check, which only checks the trace against static geometry and *some* props' collision hulls, so it doesn't take some things into account.
For the gun viewmodel I suppose we could do something like check against an approximate 2D box where the gun is on the screen.

**Comment by Graslu** [comment]
Hopefully it gets fixed in the future, even if low priority. Kind of jarring to see the lights through the gun, can't unsee after noticing it at first. 😅

**Comment by Graslu** [comment]
On Air Base the sun will go through various walls, this is due to gaps in the geometry:

![Base Profile Screenshot 2023 09 20 - 11 10 50 02](https://github.com/fgsfdsfgs/perfect_dark/assets/34986384/70ebcd8e-4fde-4083-973c-388b960fc7f2)

![Base Profile Screenshot 2023 09 20 - 11 10 41 68](https://github.com/fgsfdsfgs/perfect_dark/assets/34986384/0acdaeab-ac7e-4fdf-abca-3ce46e8ac984)

On 1964 with GLideN64 it won't occur unless you line up properly with the gap:

![1964 exe Screenshot 2023 09 20 - 11 14 28 52](https://github.com/fgsfdsfgs/perfect_dark/assets/34986384/d0c1884c-79bf-4c29-90d4-a617d3b93f91)

![1964 exe Screenshot 2023 09 20 - 11 14 57 10](https://github.com/fgsfdsfgs/perfect_dark/assets/34986384/743c96b2-64bf-40f8-967f-58100ca7873c)

On N64 this doesn't happen despite the gap being there physically.

**Comment by rafccq** [comment]
Happens in crash site as well, this is inside a cave:
![image](https://github.com/fgsfdsfgs/perfect_dark/assets/1994385/837b8cc7-dc99-4b40-83ac-80f52c97562d)

There's a seam at the right edge of the screen as well, only appears when you move the camera around:
![image](https://github.com/fgsfdsfgs/perfect_dark/assets/1994385/4751d1f9-d966-4abb-a3d9-ceb40a5a61c1)

**Comment by Graslu** [comment]
I thought characters blocked them at first, but I guess not:

![image](https://github.com/fgsfdsfgs/perfect_dark/assets/34986384/de6ddbb8-607e-4cef-8fb5-4d2ffd8c3190)

**Comment by fgsfdsfgs** [comment]
They are supposed to if they have collision. Maybe in cutscenes they don't. Or it's the result of me turning off the player collision before tracing the flares.
